src/extra/components/pulses.py

```diff
@@ -1465,25 +1465,56 @@ class DldPulses(PulsePattern):
         first_pulse_id (int, optional): Pulse ID for the first pulse,
             only used in case of missing pulse ID information in data
             and 0 by default.
+        negative_ppl_indices (bool, optional): Whether to enumerate
+            PPL-only pulses with negative indices (while keeping
+            non-negative indices for FEL-only or pumped pulses) or
+            interleave all pulses with positive indices (default).
     """
 
-    def __init__(self, detector, *, clock_ratio=None, first_pulse_id=None):
+    def __init__(self, detector, *, clock_ratio=None, first_pulse_id=None,
+                 negative_ppl_indices=False):
         super().__init__(detector, detector['raw.triggers'])
 
         self._clock_ratio = clock_ratio
         self._first_pulse_id = first_pulse_id
+        self._negative_ppl_indices = negative_ppl_indices
 
     def _get_train_ids(self):
         return np.unique(self._key.train_id_coordinates())
 
     def _get_pulse_ids(self):
         triggers = self._key.ndarray()
+        train_ids = self._key.train_id_coordinates()
 
-        index_levels = {
-            'trainId': self._key.train_id_coordinates(),
-            'pulseIndex': np.concatenate([
+        if self._negative_ppl_indices:
+            if 'ppl' not in triggers.dtype.fields:
+                raise ValueError('negative PPL indices only available '
+                                 'with PPL information')
+
+            pulse_indices = np.zeros(len(triggers), dtype=int)
+            prev_tid = -1
+
+            for i, tid, row in zip(range(len(triggers)), train_ids, triggers):
+                if tid != prev_tid:
+                    prev_tid = tid
+                    only_ppl_index = -1
+                    with_fel_index = 0
```

Comment:
There's probably a way to do this neatly with [`itertools.groupby()`](https://docs.python.org/3/library/itertools.html#itertools.groupby) to have an outer loop over train IDs. It's also not a big deal, though - I know you want to get this out quickly, and I think this code works fine.

Reply:
I tried

```python
for train_id, vals in groupby(enumerate(triggers), lambda x: train_ids[x[0]]):
    only_ppl_index = -1
    with_fel_index = 0

    for i, trigger in vals:
        if trigger['ppl'] and not trigger['fel']:
            pulse_indices[i] = only_ppl_index
            only_ppl_index -= 1
        else:
            pulse_indices[i] = with_fel_index
            with_fel_index += 1
```

but it made the overall call to `pulse_ids()` about 15% slower. While I suspect the grouping itself is faster, the nested loops seem to end up taking more time due to constructing the necessary generators.